### PR TITLE
Truncate branch name in prebuilds tab, and more

### LINKED ISSUE #5332: Truncate branch name on the prebuilds page
### Problem to solve

Branch name is wrapping when the branch name is long.

### Proposal

Truncate the branch name on the prebuilds page so that the branch name does not wrap into two lines for medium and large viewports.

| BEFORE | AFTER |
|-|-|
| branch-before | branch-after |

### PR DESCRIPTION
**Pull request by trumbitta**
## Description

https://user-images.githubusercontent.com/462705/143479443-5f0a6714-13c8-4216-bb70-bdcff31149e5.mov

This does a couple things:

- Fixes #5332 by truncating the branch name in the prebuilds tab
- Adds a title to both the branch name and the commit (which was already truncated) so that the user is still able to read the complete info even when truncated. See the attached short video.
- Aligns the branch name vertically to the rest of the info. Something that was bugging me to no end because I'm both border-line OCD and 45% UX designer 🤪. It's done with retrocompatibility in mind, to avoid a refactor. I'm willing to do the refactor if you want me to, tho.
- Updates the README with some info from the awesome @iQQBot , and a screenshot.

## Related Issue(s)

Fixes #5332

## How to test

Just start the dashboard and navigate to the prebuild tab of a project. Make sure to have the prebuild of a branch with a long name so that the actual truncation can happen.

## Release Notes

```release-note
Tweaks to the UI of the Prebuilds tab in the dashboard
Some improvements to the README of the dashboard, aimed at contributors
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **geropl** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @geropl` in a comment when ready.

Associated issue: *#5332*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by trumbitta** [comment]
@iQQBot sure thing, doing it all right now!

- Do you need me to do something about the CLA or should I wait for input by @meysholdt ?
- I chose the name README.png to keep it well sorted near the README.md :-/ is there a way for me to run the pre-commit hook before submitting the update, so that I can find the best name?
- Are you sure you want me to squash all of the commits into one? What if I squash together the 3 commits that update the prebuilds tab, then keep the one for the `ItemField` component and the one for the README apart? 3 commits total after I rebase them. (I'll do it like this without waiting for your reply, since I can always squash them all into one if you say "no, squash them all" 😄 )

**Comment by trumbitta** [comment]
Oof I fat-fingered a merge from main into this branch. Going to delete the commit right now.

**Comment by trumbitta** [comment]
No ok, I'm risking of doing more harm and it should be harmless to leave it since it's all going to go back on main either way

/werft run

👀

**Comment by meysholdt** [comment]
hi @trumbitta ! I've send you a CLA to sign via DocuSign. If you have questions, feel free to ping me via moritz@gitpod.io.

**Comment by trumbitta** [comment]
> hi @trumbitta ! I've send you a CLA to sign via DocuSign. If you have questions, feel free to ping me via [moritz@gitpod.io](mailto:moritz@gitpod.io).

Done! 🖊️

**Comment by meysholdt** [comment]
Thank you @trumbitta for signing the CLA!

File: components/dashboard/src/components/ItemsList.tsx
Comment (jankeromnes) [inline components/dashboard/src/components/ItemsList.tsx:36]: Thank you for making the vertical centering optional!

However, I see that you can now set the vertical margin via both the `className` (e.g. by setting `my-auto` and/or `my-0`) or the new `shouldCenterVertically` boolean. This creates a slight ambiguity, as future developers might accidentally provide conflicting margin classes, and then wonder why their vertical centering doesn't work.

How about shifting the responsibility of `my-auto` vs `my-0` (or no `my` class at all) to the users of `ItemField` instead? (I.e. remove `my-auto` from the className here as you did, but don't introduce a `props.shouldCenterVertically` -- simply let all the users of `ItemField` provide a `my-auto` via the `props.className` if needed).

Please let me know what you think of this alternative proposal. 🙂

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuilds.tsx:196]: Yes 🙏 thank you for that!

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuilds.tsx:202]: Also looks good to me!

One small nit: I think it's an anti-pattern to add a `truncate` class to both an element and its container. In theory, only one should be required. Maybe you can remove the `truncate` class from the container? (Haven't actually tested this -- if it doesn't work, please don't worry about this comment 🙏)

**Review by jankeromnes — COMMENTED** [review]
Hi @trumbitta, many thanks for improving Gitpod's dashboard! ✨ 🙏 

> No ok, I'm risking of doing more harm and it should be harmless to leave it since it's all going to go back on main either way

No worries here. We want to avoid merge commits on Gitpod's main branch, but since this Pull Request will get "Rebase&Merge"d eventually, this will remove the merge commit automatically.

By the way, if you wanted to remove the merge commit yourself (no need to, but just in case), you could:
- Open this Pull Request in Gitpod: https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/6907
- Run `git pull --rebase upstream main` (this will rebase your commit and remove the merge commit)
- Run `git push -f origin trumbitta/truncate-branch-name-5332` (the force-push is needed in order to change or remove commits from your branch)

---

I've had a quick look at your changes, and I must say I really like the README.md clarifications, and the use of `truncate` classes and `title` attributes. 💯 I would love to see them merged.

One minor point that is causing me to delay the approval of this Pull Request is the new `shouldCenterVertically` prop, and the slight ambiguity it creates on the CSS classes. Please see my comment inline, and let me know what you think of it (please address it if you agree, but it's also fine to simply reject my comment if you disagree).

File: components/dashboard/src/components/ItemsList.tsx
Comment (trumbitta) [reply components/dashboard/src/components/ItemsList.tsx:36]: 100% agree!

I don't know if you had the chance to take a look at my commit messages, but the gist of it is I chose to add a prop purposely for limiting the scope of my intervention and "don't go too far my first time around".

I'm totally up for a refactoring involving the 5-6 components where this one is used and removing the new prop, and I'll work on it as soon as I am able.

As a side note, there's also something that makes my spider-senses tingle about having to use both `my-auto` and `items-center` to center things, but at this moment I just don't know Tailwind nor how it's used in this codebase well enough to propose anything about it.

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (trumbitta) [reply components/dashboard/src/projects/Prebuilds.tsx:202]: Again, 100% agree.

When I was starting to work on this, I think I found an old commit adding the same thing to another piece of the app (maybe it was just a few lines above these, I can't remember) and I found the double `truncate` to be weird... but it turns out it is needed!  It doesn't work without.

That's because the CSS ellipsis needs the parent to have certain properties in order to work as expected, and the `truncate` class happens to also add those. So at the end of the day it's a hack, a workaround, that I found in the codebase and applied here.

Here's my proposal: I'll create a new `TruncatedText` component in `components/dashboard/src/components` which will:
- provide the necessary CSS "environment" for `truncate` to work
- use `truncate` on its `children` text
- accept a `title` prop

This will result in another refactoring over 38 files 😅 but, hey, I'll do it 💪.

Thoughts?
Also, since this new double `truncate` is perfectly in line with what's already around there, maybe I could open another issue / PR and work on it as my next task?

File: components/dashboard/src/components/ItemsList.tsx
Comment (trumbitta) [reply components/dashboard/src/components/ItemsList.tsx]: Refactoring done!

Comment (jankeromnes) [reply components/dashboard/src/components/ItemsList.tsx]: > Refactoring done!

Wow, many thanks! 💯 🙏 I'll take another look at your changes today.

> As a side note, there's also something that makes my spider-senses tingle about having to use both `my-auto` and `items-center` to center things, but at this moment I just don't know Tailwind nor how it's used in this codebase well enough to propose anything about it.

You're absolutely right, that's an excellent point. 🎯 From my perspective:
- centering things with flex is far superior to centering with margins (flex makes it quite easy, and the produced layout "makes sense" and works well in the majority of situations regardless of screen size, doesn't create awkward overflows, etc.)
- there is still a lot of margin-centering in the code base (or, worse, both margin- and flex-centering combined 🙈) which could be improved by removing all the auto margins and using flex instead

However, in the interest of not derailing this Pull Request too much, maybe let's focus on the `my-auto` refactoring here. We can always switch to flex in future Pull Requests. 😁

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuilds.tsx:202]: Aha, many thanks for explaining why the double `truncate` is needed @trumbitta.

Maybe the underlying reason is that Tailwind's `truncate` class is actually a combination of several CSS classes: https://tailwindcss.com/docs/text-overflow

If possible, I think I'd prefer a refactoring that makes our code always use truncate class(es) properly, instead of creating a brand new React component to solve this, but I could be wrong here.

> Also, since this new double `truncate` is perfectly in line with what's already around there, maybe I could open another issue / PR and work on it as my next task?

Agreed. 👍 I think you could indeed open an issue about unifying how we truncate text, and we could move this discussion there. Thanks for caring about this!

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuilds.tsx:185]: Aha, I think here the `my-auto` isn't actually needed (since `flex items-center` already does the vertical centering).

```suggestion
                    <ItemField className="flex items-center">
```

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuilds.tsx:194]: Same here:

```suggestion
                    <ItemField className="flex items-center">
```

File: components/dashboard/src/projects/Project.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Project.tsx:231]: Same:

```suggestion
                            <ItemField className="flex items-center">
```

File: components/dashboard/src/projects/Project.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Project.tsx:239]: Same:

```suggestion
                            <ItemField className="flex items-center">
```

File: components/dashboard/src/projects/Project.tsx
Comment (jankeromnes) [inline components/dashboard/src/projects/Project.tsx:245]: Same:

```suggestion
                            <ItemField className="flex items-center">
```

**Review by jankeromnes — APPROVED** [review]
Code looks good to me, and works like a charm. ✨ Many thanks again for improving the dashboard! 💯

I've noticed a few places where the margin-centering could be replaced by flex-centering, but this Pull Request is already a net positive as is, so I'd love to see it merged. 🚀

If you're interested and have time, I'd be more than happy to discuss / help with / review follow-up Pull Requests. 😁

/approve
/lgtm

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 6991a620a0fc6ea32cc075f2f6f8e679eb42cfb5</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6907#pullrequestreview-817600403" title="LGTM">jankeromnes</a>*

Associated issue: *#5332*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (trumbitta) [reply components/dashboard/src/projects/Prebuilds.tsx:185]: Yeah, I just went for a safe refactoring: "monkey pulls `my-auto` out of `ItemField` component, monkey adds `my-auto` to `ItemField` classes everywhere" 😄

Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuilds.tsx:185]: That's the best kind of refactoring 😁 🙏

**Pull request merged**
